### DISCUSSION TITLE
Best practices when training with incomplete data

### DISCUSSION THREAD

**Discussion by Flash-Of-Thunder** [neuralhydrology/neuralhydrology#61] [Q&A] [answered]
Hello,

I was curious if anyone had any thoughts on the best practices when training with incomplete data. Let's say I'm modeling streamflow and only 40% of my basins/gauges have full streamflow records. 

1) Is it best to just use the the 40% of full records or is it possible to utilize the incomplete sets to improve the model? E.g. should I interpolate missing values or replace them with the median if only a handful of values are missing. 

2) How does NH handle missing values in the training set by default? Right now I'm combining daymet climate data with USGS streamflow values model flow. Daymet is a 365-calendar day dataset that removes the extra day in leap years, so when I merge it with full streamflow data I have NAs ever 4 years. Is this inconsequential in the model? Would it be better to interpolate the Daymet to fill NAs?

If this has already been answered apologies, and I'd be happy to read a link. Cheers!

**Comment by kratzert** [comment, marked as answer]
Good question and I see that we lack documentation here:

In NeuralHydrology, samples where the _target_ is NaN are ignored, when computing the loss during training but will get a prediction during validation/testing. Samples, where any input value is NaN are removed. The reason is that a single NaN value in the inputs will lead to all-NaNs in the model outputs.

So to your questions:

1: Always better to use all data and nothing to worry here from your side. The code will automatically ignore all samples where the target value is missing.

2: DayMet is problematic, as NaN in the input data is more of a problem than NaN in the targets. In CAMELS, the 31st of December in leap years is removed (by the authors of CAMELS not us!). So it seems like the dataset is continuous without missing data, even though that in leap years you are missing one day. You could probably do the same, or interpolate the inputs. The issue here is that if you have a NaN in the inputs, all input sequences to the LSTM, where this date is used as input (i.e. for the next n days where n is your sequence length defined by `seq_length` e.g. 365) will be removed. This is a lot of thrown-away data eventually and it might be better to find another solution for the missing inputs.

**Reply by kratzert** [reply]
Here is the function that decides if an input sample is valid (can be used as input) or not https://github.com/neuralhydrology/neuralhydrology/blob/31bd284c4713dce1404818f9416161303c25b436/neuralhydrology/datasetzoo/basedataset.py#L653

**Reply by Flash-Of-Thunder** [reply]
This is exceptionally helpful, thanks!

**Comment by danklotz** [comment]
This does not help you directly, but on a conceptual level you might also find [this paper](https://hess.copernicus.org/preprints/hess-2021-515/)  relevant.
